src/main/java/com/nvs/store/service/CartService.java

```diff
@@ -0,0 +1,202 @@
+            }
+        }
+
+        validateAvailableProductQuantities(product, cartItemRequest.getQuantity());
+        CartItem cartItem = CartItem
+                .builder()
+                .quantity(cartItemRequest.getQuantity())
+                .product(product)
+                .cart(cart)
+                .build();
+
+        cartItemRepository.save(cartItem);
+
+        return CartUtils.convertToCartDto
+                .apply(cartRepository
+                        .findByUserIdAndActualTrue(user.getId()));
+    }
+
+    public CartDTO orderCart() {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository.findByEmail(userEmail).orElseThrow(
+                () -> new RuntimeException(
+                        String.format("User with username %s not found", userEmail))
+        );
+
+        Cart cart = cartRepository
+                .findByUserIdAndActualTrue(user.getId());
+
+        List<CartItem> orderedCartItems = cartItemRepository.findCartItemByCart(cart);
+
+        for (CartItem cartItem : orderedCartItems) {
+            Product product = productRepository.findById(cartItem.getProduct().getId())
+                    .orElseThrow(() -> new RuntimeException(
+                            String.format("Product with id [%d] does not exist", cartItem.getProduct().getId()))
+                    );
+            if (product.getAvailable() < 1) {
+                throw new RuntimeException(
+                        String.format("Sorry, [%s] is not available yet", product.getTitle())
+                );
+            } else if (cartItem.getQuantity() > product.getAvailable()) {
+                throw new RuntimeException(
+                        String.format(
+                                "Only [%d] of [%s] available. Please order available quantity.",
+                                product.getAvailable(),
+                                product.getTitle())
+                );
+            }
+        }
+
+        cart.setActual(false);
+        cartRepository.save(cart);
+
+        Cart newCart = Cart.builder()
+                .user(user)
+                .actual(true)
+                .cartItems(List.of())
+                .build();
+
+        return CartUtils.convertToCartDto
+                .apply(cartRepository
+                        .save(newCart));
+    }
+
+
+    public CartDTO updateProduct(Long id, Integer quantity) {
+        String userEmail = getUserEmailFromContext();
+        User user = userRepository.findByEmail(userEmail).orElseThrow(
+                () -> new RuntimeException(
+                        String.format("User with username %s not found", userEmail))
+        );
+        CartItem cartItem = cartItemRepository.findCartItemByProductId(id)
+                .orElseThrow(() -> new RuntimeException("No such product in cart"));
+
+        CartItemRequest cartItemRequest = CartItemRequest.builder()
```

Comment:
Why do you need to initiate a new object of type CartItemRequest?

Reply:
yeah , stupid move 😅
fixed ✅